proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,402 @@
+# Choosing Strings
+
+## Preamble
+This note is part of a series of notes that gives complete end-to-end examples
+of using Interface Types to export and import functions.
+
+## Introduction
+
+Keeping track of memory that is allocated in order to allow string processing
+can be challenging. This example illustrates an extreme case that involves the
+non-deterministic computation of string values.
+
+This scenario is a model of many situations where the actual flow of data
+between a call and callee can be very difficult to predict. A consequence of
+that is that one cannot easily predict when given allocated blocks may be freed.
+
+Our exemplar for this is the non-deterministic C++ chooser function; which
+(might) return one of its `string` arguments as its return value:
+
+```C++
+typedef std::shared_ptr<std::string> shared_string;
+
+shared_string nondeterministic_choice(shared_string one, shared_string two) {
+  return random() > 0.5 ? std::move(one) : std_move(two);
+}
+```
+
+This function takes two string arguments and returns one of them -- at
+random. It uses C++ `shared_ptr` structures to try to avoid any unnecessary
+copying; whilst also potentially allowing the callee to keep hold of its input
+arguments.
+
+Regardless of the merits of this particular function, it sets up significant
+challenges should we want to expose it using Interface Types. Specifically,
+there are two resources entering the function, with just one leaving. However,
+when exposed as an Interface Type function, all these resources must be created
+and properly disposed of within the adapter code itself.
+
+## Exporting the Chooser
+
+The Interface Type function signature for our chooser is simple: the function
+takes two `string`s and returns one:
+
+```wasm
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+)
+```
+
+One of the specific challenges in this scenario is the handling of shared
+pointers. We 'require' that the core `nondeterministic_choice` function honor
+the semantics of proper reference counting the input arguments and the returned
+string.
+
+However, there are actually _two_ schemes in play in this scenario: the
+Interface Types management of resources and the C++ implementation of
+`shared_ptr`. Effectively, in addition to lifting and lowering the `string`
+values we must also lift and lower the ownership between Interface Types and
+core WASM.
+
+In this scenario we take a somewhat simplified view of C++'s implementation of
+shared pointers: a shared pointer is implemented as a pair consisting of a
+reference count and a raw pointer to the shared resource.
+
+>Note: In practice, the C++ implementation of shared pointers is somewhat more
+>complex; for reasons that are not important here.
+
+This results in _two_ memory allocations for the shared resource: one for the
```

Comment:
It's probably a little late for this (since I think this code does work), but did you consider passing the shared pointers as pairs instead of allocated structs? That would have reduced the number of allocations in the example, and (to my reading) made it more "realistic" by passing the refcounts by value instead of by reference (as is implicit in the example code).

Reply:
Ah. Was trying to emulate a simplified version of shared_ptr. Cant pass refcounts themselves by value; (suppose callee decides to take an extra reference to a string?)

Reply:
When I woke up this morning I'd realized where I went awry: while "real" shared_ptrs themselves are passed-by-value, they share an underlying auxiliary structure. Whereas in your case the shared_ptr itself is that structure. That makes the ABI for the function rather surprising for something with passed-by-value parameters, but in the end it amounts to the same thing.